Frameworks/AddressBook/ABAddressBookManagerInternal.mm

```diff
@@ -82,4 +189,171 @@ - (NSArray*)getListOfContacts {
     return result;
 }
 
+- (NSArray*)getListOfModifiableContacts {
+    NSMutableArray* result = [[NSMutableArray alloc] init];
+    WACContactReader* reader = [_contactList getContactReader];
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    __block BOOL shouldContinue = YES;
+    while (shouldContinue) {
+        // Keep reading batches until the batch is empty.
+        [reader readBatchAsyncWithSuccess:^(WACContactBatch* success) {
+            if ([success.contacts count] == 0) {
+                shouldContinue = NO;
+            } else {
+                shouldContinue = YES;
+                for (WACContact* contact in success.contacts) {
+                    _ABContact* wrappedContact = [[_ABContact alloc] initWithContact:contact andType:kAddressBookReadWriteContact];
+                    wrappedContact.manager = (__bridge ABRecordRef)self;
+                    [result addObject:wrappedContact];
+                }
+            }
+            dispatch_semaphore_signal(semaphore);
+        }
+
+            failure:^(NSError* error) {
+                shouldContinue = NO;
+                dispatch_semaphore_signal(semaphore);
+            }];
+
+        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+    }
+
+    dispatch_release(semaphore);
+    return result;
+}
+
+- (bool)addContact:(ABRecordRef)record error:(CFErrorRef*)error {
+    _ABContact* contact = (__bridge _ABContact*)record;
+
+    switch (contact.type) {
+        case kAddressBookNewContact: {
+            // New contacts can't be added twice.
+            if ([_toAdd containsObject:contact.contact]) {
```

Comment:
Ah, I see. Below.

---
In reply to: [77099735](https://github.com/Microsoft/WinObjC/pull/935#discussion_r77099735) [](ancestors = 77099735)